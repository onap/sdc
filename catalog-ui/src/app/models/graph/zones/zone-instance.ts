import {PolicyInstance} from "app/models/graph/zones/policy-instance";
import {GroupInstance} from "./group-instance";
import {Component as TopologyTemplate} from "app/models";
import {IUiBaseObject} from "../../ui-models/ui-base-object";
import { Subject } from "rxjs";

export enum ZoneInstanceMode {
    NONE,
    HOVER,
    SELECTED,
    TAG
}

export enum ZoneInstanceType {
    GROUP,
    POLICY
}

export enum ZoneInstanceAssignmentType {
    COMPONENT_INSTANCES,
    GROUPS
}

export interface IZoneInstanceMethod {

    getSavedAssignments():Array<IZoneInstanceAssignment>;
    setSavedAssignments(newAssignments:Array<IZoneInstanceAssignment>):void;
}

export interface IZoneInstanceAssignment extends IUiBaseObject{
    type: ZoneInstanceAssignmentType
}

export class ZoneInstance {

    parentComponentType:string;
    parentComponentID:string;
    instanceData: PolicyInstance | GroupInstance;
    mode:ZoneInstanceMode;
    type:ZoneInstanceType;
    handle:string;
    assignments:Array<IZoneInstanceAssignment>; //temp assignments visible on the UI; not the saved values on the BE
    hidden:boolean;
    forceSave:Subject<Function>;

    constructor(instance: PolicyInstance | GroupInstance, topologyTemplateType: string, topologyTemplateId: string) {

        this.instanceData = instance;
        this.parentComponentType = topologyTemplateType;
        this.parentComponentID = topologyTemplateId;

        if (instance instanceof PolicyInstance) {
            this.type = ZoneInstanceType.POLICY;
        } else {
            this.type = ZoneInstanceType.GROUP;
        }

        this.assignments = this.instanceData.getSavedAssignments();
        this.mode = ZoneInstanceMode.NONE;
        this.hidden = false;
        this.forceSave = new Subject();
    }

    public isAlreadyAssigned = (nodeId:string):boolean => {
        let matchingAssignments = this.assignments.filter((assignment) => {
            return assignment.uniqueId == nodeId;
        });
        return matchingAssignments && matchingAssignments.length > 0;
    }

    public addOrRemoveAssignment = (nodeId:string, nodeType:ZoneInstanceAssignmentType)=> { //change temp assignments, unsaved but visible in UI.

        if (!this.isAlreadyAssigned(nodeId)) {
            this.assignments.push(<IZoneInstanceAssignment>{uniqueId: nodeId, type: nodeType});
        } else {
            this.assignments = this.assignments.filter(assignment => assignment.uniqueId != nodeId);
        }
    }
    
    public isZoneAssignmentChanged(oldAssignments:Array<IZoneInstanceAssignment>, newAssignments:Array<IZoneInstanceAssignment>):boolean {
        if (oldAssignments.length != newAssignments.length) {
            return true;
        }
        let difference:Array<IZoneInstanceAssignment> = oldAssignments.filter((oldAssignment) => {
            return !newAssignments.find(newAssignment => newAssignment.uniqueId == oldAssignment.uniqueId);
        });
        if (difference.length) {
            return true;
        }

        return false;
    }

    public updateInstanceData (instanceData:  PolicyInstance | GroupInstance):void {
        this.instanceData = instanceData;
        this.assignments = this.instanceData.getSavedAssignments();
    }

    public showHandle = (handleId:string) => {
        this.handle = handleId;
    }

    public hideHandle = ():void => {
        this.handle = null;
    }
}
