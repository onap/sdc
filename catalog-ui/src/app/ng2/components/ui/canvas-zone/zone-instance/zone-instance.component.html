<div class="zone-child mode-{{config.mode}}" [class.locked]="activeInstanceMode > MODE.HOVER" 
     (mouseenter)="setMode(MODE.HOVER)" (mouseleave)="setMode(MODE.NONE)" (click)="setMode(MODE.SELECTED)">
        <div class="zone-child__handle" (click)="setMode(MODE.TAG, $event)">+</div>
        <div class="zone-child__body">
            <div class="zone-child__body-content">{{config.count || defaultIconText}}</div>
        </div>
        <div class="zone-child__name">{{config.name}}</div>
</div>