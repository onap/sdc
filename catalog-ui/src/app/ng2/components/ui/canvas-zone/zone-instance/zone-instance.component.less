.zone-child {
    position:relative;
    width:76px;
    margin:5px;

    .zone-child__handle {
        display:none;
        position:absolute;
        right:4px;
        top:10px;
        width:22px;
        height:22px;
        cursor:pointer;
        border: solid white 1px;
        border-radius: 2px;
        text-align: center;
        font-weight:bold;
    }

    .zone-child__body {
        margin:0 auto;
        width:43px;
        height:43px;
        display:flex;
        padding:3px;

    }

    .zone-child__body-content {
        border-radius: 2px;
        flex:1;
        color:white;
        font-size:18px;
        text-align:center;       
        display:flex;
        align-items: center;
        justify-content: center;
    }

    .zone-child__name {
        text-overflow: ellipsis;
        overflow: hidden;
        white-space: nowrap;
        text-align:center;
    }
    /* Dynamic classes below */

    &.mode-1, &.mode-2, &.mode-3 { //hover, selected, tag
        .zone-child__body {
            border:solid 2px;
            border-radius: 2px;
            padding:2px;
            cursor:pointer;
        }
        .zone-child__handle{
            display:block;
            cursor:pointer;
        }
    }

    &.locked {
        cursor: default;
    }
    
    // &:not(.locked):hover .zone-child__handle{
    //     display:block;
    // }
    .zone-child__body {
        cursor: default;
    }
    &.mode-3 .zone-child__handle {
        width:24px;
        height:24px;
        right:3px;
        top:9px;
        display:block;
        background-image: linear-gradient(-140deg, #009E98 0%, #97D648 100%);
        border: 2px solid #FFFFFF;
        border-radius: 2px;
        box-shadow: inset 2px -2px 3px 0 #007A3E;
        cursor: pointer;
    }

}
.sdc-canvas-zone.group-zone {
    .zone-child__handle {
        background-color:#009FDB;
    }
    .zone-child__body {
        border-color:#009FDB;
    }
    .zone-child__body-content {
        background: #009FDB;
    }
}

.sdc-canvas-zone.policy-zone {
    .zone-child__handle {
        background-color:#0568AE;
    }
    .zone-child__body {
        border-color:#1287D9;
        .zone-child__body-content {
            background: #1287D9;
        }
    }
    .zone-child__body-content {
        background: #0568AE;
    }
}