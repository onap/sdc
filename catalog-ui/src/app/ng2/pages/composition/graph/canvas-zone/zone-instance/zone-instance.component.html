<!--
 ~ Copyright (C) 2018 AT&T Intellectual Property. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 -->


<div #currentComponent class="zone-instance mode-{{zoneInstance.mode}}" [class.locked]="activeInstanceMode > MODE.HOVER"
     [class.hiding]="hidden"
     (mouseenter)="setMode(MODE.HOVER)" (mouseleave)="setMode(MODE.NONE)" (click)="setMode(MODE.SELECTED, $event)">
    <div class="zone-instance__body" sdc-tooltip tooltip-text="{{zoneInstance.instanceData.name}}" [attr.data-tests-id]="zoneInstance.instanceData.name">
        <div *ngIf="zoneInstance.handle" class="target-handle {{zoneInstance.handle}}"
             (click)="tagHandleClicked($event)"></div>
        <div *ngIf="!isViewOnly" class="zone-instance__handle" (click)="setMode(MODE.TAG, $event)">+</div>
        <div class="zone-instance__body-content">{{zoneInstance.assignments.length || defaultIconText}}</div>
    </div>
    <div class="zone-instance__name">{{zoneInstance.instanceData.name}}</div>
</div>
