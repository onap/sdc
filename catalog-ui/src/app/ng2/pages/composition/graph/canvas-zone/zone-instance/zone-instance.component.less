@import '../../../../../../../assets/styles/variables';

.zone-instance {
    
    width:76px;
    margin:5px;
    opacity:1;

    .zone-instance__handle {
        display:none;
        position:absolute;
        left: 31px;
        top: 8px;
        width:22px;
        height:22px;
        cursor:pointer;
        border: solid @main_color_p 1px;
        border-radius: 2px;
        text-align: center;
        font-weight:bold;
    }

    .zone-instance__body {
        position:relative;
        margin:0 auto;
        width:43px;
        height:43px;
        display:flex;
        padding:3px;
    }

    .zone-instance__body-content {
        border-radius: 2px;
        flex:1;
        color:@main_color_p;
        font-size:16px;
        text-align:center;       
        display:flex;
        align-items: center;
        justify-content: center;
        box-shadow:none;
        transition:box-shadow 5s;
    }

    .zone-instance__name {
        text-overflow: ellipsis;
        overflow: hidden;
        white-space: nowrap;
        text-align:center;
    }
    /* Dynamic classes below */

    .target-handle {
        position:absolute;
        width:18px;
        height:18px;
        display:block;
        top: -4px;
        right: -6px;
        background-size: 100% 100%;
        cursor: url("../../../../../../../assets/styles/images/canvas-tagging-icons/policy_2.svg"), pointer;

        &.tagged-policy {
            background-image: url('../../../../../../../assets/styles/images/canvas-tagging-icons/policy_added.svg');
        }
    
        &.tag-available {
            background-image: url('../../../../../../../assets/styles/images/canvas-tagging-icons/indication.svg');
        }
    }


    &.mode-1, &.mode-2, &.mode-3 { //hover, selected, tag
        .zone-instance__body {
            border:solid 2px;
            border-radius: 2px;
            padding:2px;
            cursor:pointer;
        }
    }

    &.mode-1, &.mode-2:hover{
        .zone-instance__handle{
            display:block;
        }
    }

    &.locked {
        cursor: inherit;
    }

    &.hiding {
        opacity:0;
        .zone-instance__body-content {
            box-shadow: #CCC 0px 0px 15px;
        }
    }
    

    &.mode-3 .zone-instance__handle {
        width:24px;
        height:24px;
        right:-6px;
        top:7px;
        display:block;
        background-image: linear-gradient(-140deg, #009E98 0%, #97D648 100%);
        border: 2px solid @main_color_p;
        border-radius: 2px;
        box-shadow: inset 2px -2px 3px 0 #007A3E;
    }

}
.sdc-canvas-zone.group-zone {
    .zone-instance__handle {
        background-color:@main_color_a;
    }
    .zone-instance__body {
        border-color:@main_color_a;
        .zone-instance__body-content {
            background: @main_color_a;
        }
    }
}

.sdc-canvas-zone.policy-zone {
    .zone-instance__handle {
        background-color:@main_color_r;
    }
    .zone-instance__body {
        border-color:@main_color_r;
        .zone-instance__body-content {
            background: @main_color_r;
        }
    }
}