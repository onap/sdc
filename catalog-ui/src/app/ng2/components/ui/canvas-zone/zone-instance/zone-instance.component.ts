import { Component, Input, Output, EventEmitter, ViewEncapsulation } from '@angular/core';
import { ZoneInstanceConfig, ZoneInstanceMode } from 'app/models/graph/zones/zone-child';

@Component({
    selector: 'zone-instance',
    templateUrl: './zone-instance.component.html',
    styleUrls: ['./zone-instance.component.less'],
    encapsulation: ViewEncapsulation.None
})
export class ZoneInstanceComponent {

    @Input() config:ZoneInstanceConfig;
    @Input() defaultIconText:string;
    @Input() isActive:boolean;
    @Input() activeInstanceMode: ZoneInstanceMode;
    @Output() modeChange: EventEmitter<any> = new EventEmitter<any>();
    private MODE = ZoneInstanceMode;

    private setMode = (mode:ZoneInstanceMode, event?:any):void => {
        if(!this.isActive || this.isActive && mode == ZoneInstanceMode.TAG){ //when active, do not allow hover/select mode toggling
            this.modeChange.emit({newMode: mode, instance: this.config});
        }
        if(event){
            event.stopPropagation();
        }
    }

}