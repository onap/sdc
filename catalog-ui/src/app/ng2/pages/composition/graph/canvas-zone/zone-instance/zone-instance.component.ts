import { Component, Input, Output, EventEmitter, ViewEncapsulation, OnInit, SimpleChange, ElementRef, ViewChild, SimpleChanges } from '@angular/core';
import {
    ZoneInstance, ZoneInstanceMode, ZoneInstanceType,
    IZoneInstanceAssignment
} from 'app/models/graph/zones/zone-instance';
import { PoliciesService } from 'app/ng2/services/policies.service';
import { GroupsService } from 'app/ng2/services/groups.service';
import { IZoneService } from "app/models/graph/zones/zone";
import { EventListenerService } from 'app/services';
import { GRAPH_EVENTS } from 'app/utils';
import { Subject } from 'rxjs';
import { Store } from "@ngxs/store";
import { CompositionService } from "app/ng2/pages/composition/composition.service";
import { PolicyInstance } from "app/models";
import {SelectedComponentType, SetSelectedComponentAction} from "../../../common/store/graph.actions";


@Component({
    selector: 'zone-instance',
    templateUrl: './zone-instance.component.html',
    styleUrls: ['./zone-instance.component.less'],
    encapsulation: ViewEncapsulation.None
})
export class ZoneInstanceComponent implements OnInit {

    @Input() zoneInstance:ZoneInstance;
    @Input() defaultIconText:string;
    @Input() isActive:boolean;
    @Input() isViewOnly:boolean;
    @Input() activeInstanceMode: ZoneInstanceMode;
    @Input() hidden:boolean;
    @Input() forceSave:Subject<Function>;
    @Output() modeChange: EventEmitter<any> = new EventEmitter<any>();
    @Output() assignmentSaveStart: EventEmitter<void> = new EventEmitter<void>();
    @Output() assignmentSaveComplete: EventEmitter<boolean> = new EventEmitter<boolean>();
    @Output() tagHandleClick: EventEmitter<ZoneInstance> = new EventEmitter<ZoneInstance>();
    @ViewChild('currentComponent') currentComponent: ElementRef;
    private MODE = ZoneInstanceMode;
    private zoneService:IZoneService;

    constructor(private policiesService:PoliciesService, private groupsService:GroupsService, private eventListenerService:EventListenerService, private compositionService:CompositionService, private store:Store){}

    ngOnInit(){
        if(this.zoneInstance.type == ZoneInstanceType.POLICY){
            this.zoneService = this.policiesService;
        } else {
            this.zoneService = this.groupsService;
        }
        if(this.forceSave) {
            this.forceSave.subscribe((afterSaveFunction:Function) => {
                this.setMode(ZoneInstanceMode.TAG, null, afterSaveFunction);
            })
        }
    }

    ngOnChanges(changes:SimpleChanges) {
        if(changes.hidden){
            this.currentComponent.nativeElement.scrollIntoView({behavior: "smooth", block: "nearest", inline:"end"});
        }
    }

    ngOnDestroy() {
        if(this.forceSave) {
            this.forceSave.unsubscribe();
        }
    }

    private setMode = (mode:ZoneInstanceMode, event?:any, afterSaveCallback?:Function):void => {
        
        if(event){ //prevent event from handle and then repeat event from zone instance
            event.stopPropagation();
        }

        if(!this.isActive && this.activeInstanceMode === ZoneInstanceMode.TAG) {
            return; //someone else is tagging. No events allowed
        }

        if(this.isActive && this.zoneInstance.mode === ZoneInstanceMode.TAG){
            if(mode !== ZoneInstanceMode.TAG) {
                return; //ignore all other events. The only valid option is saving changes.
            }

            let oldAssignments:Array<IZoneInstanceAssignment> = this.zoneInstance.instanceData.getSavedAssignments();
            if(this.zoneInstance.isZoneAssignmentChanged(oldAssignments, this.zoneInstance.assignments)) {

                this.assignmentSaveStart.emit();
                
                this.zoneService.updateZoneInstanceAssignments(this.zoneInstance.parentComponentType, this.zoneInstance.parentComponentID, this.zoneInstance.instanceData.uniqueId, this.zoneInstance.assignments).subscribe(
                    (success) => {
                        this.zoneInstance.instanceData.setSavedAssignments(this.zoneInstance.assignments);
                        
                        if(this.zoneInstance.instanceData instanceof PolicyInstance){
                            this.compositionService.updatePolicy(this.zoneInstance.instanceData);
                            this.eventListenerService.notifyObservers(GRAPH_EVENTS.ON_POLICY_INSTANCE_UPDATE, this.zoneInstance.instanceData);
                            this.store.dispatch(new SetSelectedComponentAction({component: this.zoneInstance.instanceData, type: SelectedComponentType.POLICY}));
                        } else {
                            this.compositionService.updateGroup(this.zoneInstance.instanceData);
                            this.eventListenerService.notifyObservers(GRAPH_EVENTS.ON_GROUP_INSTANCE_UPDATE, this.zoneInstance.instanceData);
                            this.store.dispatch(new SetSelectedComponentAction({component: this.zoneInstance.instanceData, type: SelectedComponentType.GROUP}));
                        }

                        this.assignmentSaveComplete.emit(true);
                        if(afterSaveCallback) afterSaveCallback();
                    }, (error) => {
                        this.zoneInstance.assignments = oldAssignments;
                        this.assignmentSaveComplete.emit(false);
                });
            } else {
                if(afterSaveCallback) afterSaveCallback();
            }
            this.modeChange.emit({newMode: ZoneInstanceMode.NONE, instance: this.zoneInstance});
            // this.store.dispatch(new unsavedChangesActions.RemoveUnsavedChange(this.zoneInstance.instanceData.uniqueId));


        } else {
            this.modeChange.emit({newMode: mode, instance: this.zoneInstance});
            if(mode == ZoneInstanceMode.TAG){
                // this.store.dispatch(new unsavedChangesActions.AddUnsavedChange(this.zoneInstance.instanceData.uniqueId));
            }
        }         
    } 

    private tagHandleClicked = (event:Event) => {
        this.tagHandleClick.emit(this.zoneInstance);
        event.stopPropagation();
    };

}