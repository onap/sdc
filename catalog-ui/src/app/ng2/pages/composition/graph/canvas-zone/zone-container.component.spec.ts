import {async, ComponentFixture} from '@angular/core/testing';
import {By} from '@angular/platform-browser';
import {ConfigureFn, configureTests} from '../../../../../../jest/test-config.helper';
import 'jest-dom/extend-expect';
import {ZoneInstanceType} from '../../../../../../app/models/graph/zones/zone-instance';
import {ZoneContainerComponent} from './zone-container.component';


describe('ZoneContainerComponent', () => {
    let fixture: ComponentFixture<ZoneContainerComponent>;

    beforeEach(
        async(() => {
            const configure: ConfigureFn = testBed => {
                testBed.configureTestingModule({
                    declarations: [ZoneContainerComponent]
                });
            };

            configureTests(configure).then(testBed => {
                fixture = testBed.createComponent(ZoneContainerComponent);
            });
        })
    );


    it('should match current snapshot of palette element component', () => {
        expect(fixture).toMatchSnapshot();
    });

    it('should have a group-zone class when the ZoneInstanceType is GROUP',
        () => {
            fixture.componentInstance.type = ZoneInstanceType.GROUP;
            fixture.detectChanges();
            const compiled = fixture.debugElement.query(By.css('.sdc-canvas-zone'));
            expect(compiled.nativeElement).toHaveClass('group-zone');
        });

    it('should have a policy-zone class when the ZoneInstanceType is POLICY',
        () => {
            fixture.componentInstance.type = ZoneInstanceType.POLICY;
            fixture.detectChanges();
            const compiled = fixture.debugElement.query(By.css('.sdc-canvas-zone'));
            expect(compiled.nativeElement).toHaveClass('policy-zone');
        });
});