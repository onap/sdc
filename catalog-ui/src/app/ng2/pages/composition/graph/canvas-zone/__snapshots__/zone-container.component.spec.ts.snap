// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZoneContainerComponent should match current snapshot of palette element component 1`] = `
<zone-container
  backgroundClick={[Function EventEmitter]}
  backgroundClicked={[Function Function]}
  minimize={[Function EventEmitter]}
  unminifyZone={[Function Function]}
>
  <div>
    <div
      class="sdc-canvas-zone__header"
    >
      <div
        class="sdc-canvas-zone__title"
      >
        
        <span
          class="sdc-canvas-zone__counter"
        >
          
        </span>
      </div>
      <span
        class="sdc-canvas-zone__state-button"
      >
        â€“
      </span>
    </div>
    <div
      class="sdc-canvas-zone__container"
    />
  </div>
</zone-container>
`;
